homeassistant/components/devolo_home_network/binary_sensor.py

```diff
@@ -0,0 +1,123 @@
+"""Platform for binary sensor integration."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+
+from devolo_plc_api.device import Device
+
+from homeassistant.components.binary_sensor import (
+    DEVICE_CLASS_PLUG,
+    DEVICE_CLASS_UPDATE,
+    BinarySensorEntity,
+    BinarySensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ENTITY_CATEGORY_DIAGNOSTIC
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import (
+    CONNECTED_PLC_DEVICES,
+    CONNECTED_TO_ROUTER,
+    DOMAIN,
+    FIRMWARE_UPDATE_AVAILABLE,
+)
+from .entity import DevoloEntity
+
+
+def _is_connected_to_router(entity: DevoloBinarySensorEntity) -> bool:
+    """Check, if device is attached to the router."""
+    return all(
+        device["attached_to_router"]
+        for device in entity.coordinator.data["network"]["devices"]
+        if device["mac_address"] == entity.device.mac
+    )
+
+
+@dataclass
+class DevoloBinarySensorRequiredKeysMixin:
+    """Mixin for required keys."""
+
+    value_func: Callable[[DevoloBinarySensorEntity], bool]
+
+
+@dataclass
+class DevoloBinarySensorEntityDescription(
+    BinarySensorEntityDescription, DevoloBinarySensorRequiredKeysMixin
+):
+    """Describes devolo sensor entity."""
+
+
+SENSOR_TYPES: dict[str, DevoloBinarySensorEntityDescription] = {
+    CONNECTED_TO_ROUTER: DevoloBinarySensorEntityDescription(
+        key=CONNECTED_TO_ROUTER,
+        device_class=DEVICE_CLASS_PLUG,
+        entity_category=ENTITY_CATEGORY_DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+        icon="mdi:router-network",
+        name="Connected to router",
+        value_func=_is_connected_to_router,
+    ),
+    FIRMWARE_UPDATE_AVAILABLE: DevoloBinarySensorEntityDescription(
+        key=FIRMWARE_UPDATE_AVAILABLE,
+        device_class=DEVICE_CLASS_UPDATE,
+        entity_category=ENTITY_CATEGORY_DIAGNOSTIC,
+        entity_registry_enabled_default=True,
+        icon="mdi:update",
+        name="Firmware update available",
+        value_func=lambda entity: entity.coordinator.data["result"] == "UPDATE_AVAILABLE",  # type: ignore[no-any-return]
```

Comment:
In order to avoid the type ignore you might be able to also use a separate function `_is_update_available`.

Reply:
Interestingly that doesn't help. It just moves the place, where I need the ignore. I assume, I would need to specify the coordinator data more precise, but as that is JSON, that could turn out ugly ...

Reply:
It would be nice to get this typed in the upstream lib and return a dataclass instead of JSON, but thats probably too big of a change here.

Reply:
I like the idea of a dataclass and will look into it. But I agree that this should be a separate step.